gio/src/cancellable_future.rs

```diff
@@ -0,0 +1,143 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::cancellable::CancellableExtManual;
+use crate::prelude::CancellableExt;
+use crate::Cancellable;
+use crate::IOErrorEnum;
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+pin_project! {
+    /// A future which can be cancelled via [`Cancellable`].
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let c = Cancellable::new();
+    /// ctx.spawn_local(CancellableFuture::new(task, c)
+    ///                            .await
+    ///                            .map(|_| ()))
+    ///
+    /// ```
+    pub struct CancellableFuture<F> {
+        #[pin]
+        future: F,
+        cancellable: Cancellable,
+    }
+}
+
+impl<F> CancellableFuture<F> {
+    /// Creates a new `CancellableFuture` using a [`Cancellable`].
+    ///
+    /// When [`cancel`](CancellableExt::cancel) is called, the future will complete
+    /// immediately without making any further progress. In such a case, an error
+    /// will be returned by this future (i.e., [`IOErrorEnum::Cancelled`]).
+    pub fn new(future: F, cancellable: Cancellable) -> Self {
+        Self {
+            future,
+            cancellable,
+        }
+    }
+
+    /// Checks whether the future has been cancelled.
+    ///
+    /// Note that all this method indicates is whether [`cancel`](CancellableExt::cancel)
+    /// was called. This means that it will return true even if:
+    ///   * `cancel` was called after the future had completed.
+    ///   * `cancel` was called while the future was being polled.
+    #[inline]
+    pub fn is_cancelled(&self) -> bool {
+        self.cancellable.is_cancelled()
+    }
+}
+
+impl<F> Future for CancellableFuture<F>
+where
+    F: Future,
+{
+    type Output = Result<<F as Future>::Output, glib::Error>;
```

Comment:
As the only possible error here is `Cancelled`, maybe it should get its own error type

Reply:
Done. I've added a conversion trait for `glib::Error` tho. I think it might nice to have a integration with the rest of the library

Reply:
Sounds like a good idea, yes